src/formula.rs

```diff
@@ -350,6 +356,27 @@ impl<'a> FormulaRef<'a> {
         f
     }
 
+    /// Returns a copy of this formula with a new root
+    /// (in the same tree)
+    fn to_rerooted(self, new_root: NodeId) -> Formula {
+        Formula {
+            root: new_root,
+            tree: self.formula.tree.clone(),
+            typecode: self.compute_typecode_at(new_root),
+            variables: self.formula.variables.clone(),
+        }
+    }
+
+    /// Computes the typecode of the given node
+    /// according to the corresponding statement
+    fn compute_typecode_at(&self, node_id: NodeId) -> TypeCode {
+        let sref = self
+            .db
+            .statement_by_label(self.formula.tree[node_id])
+            .unwrap();
```

Comment:
change to `expect` and describe validity assumptions (is this impossible by construction of the database, or is this a user-facing precondition?)

Reply:
Sorry I was too fast merging!
Agreed and well noted, this will come into the next one.

Formulas can be obtained either from `Grammar::parse_formula` for a custom string of tokens, or from `StmtParse::get_formula` for statements of the database. In both cases, only valid database labels are used.